src/main/java/com/google/cloud/genomics/dataflow/functions/JoinGvcfVariants.java

```diff
@@ -0,0 +1,189 @@
+ * or implied. See the License for the specific language governing permissions and limitations under
+ * the License.
+ */
+package com.google.cloud.genomics.dataflow.functions;
+
+import static com.google.common.collect.Lists.newArrayList;
+
+import java.util.Collections;
+import java.util.Comparator;
+import java.util.Iterator;
+import java.util.LinkedList;
+import java.util.List;
+
+import com.google.api.client.util.Lists;
+import com.google.api.services.genomics.model.SearchVariantsRequest;
+import com.google.api.services.genomics.model.Variant;
+import com.google.cloud.dataflow.sdk.transforms.DoFn;
+import com.google.cloud.dataflow.sdk.transforms.GroupByKey;
+import com.google.cloud.dataflow.sdk.transforms.ParDo;
+import com.google.cloud.dataflow.sdk.values.KV;
+import com.google.cloud.dataflow.sdk.values.PCollection;
+import com.google.cloud.genomics.dataflow.readers.VariantReader;
+import com.google.cloud.genomics.dataflow.utils.GenomicsDatasetOptions;
+import com.google.cloud.genomics.dataflow.utils.VariantUtils;
+import com.google.cloud.genomics.utils.GenomicsFactory;
+import com.google.common.base.Preconditions;
+
+/**
+ * The transforms in this class convert data in Genome VCF (gVCF) format to variant-only VCF data
+ * with calls from block-records merged into the variants with which they overlap.
+ * 
+ * This is currently done only for SNP variants. Indels and structural variants are left as-is.
+ */
+public class JoinGvcfVariants {
+
+  public static final String GVCF_VARIANT_FIELDS =
+      "nextPageToken,variants(referenceName,start,end,referenceBases,alternateBases,calls(genotype,callSetName))";
+
+  private static final List<String> REQUIRED_FIELDS = newArrayList("nextPageToken",
+      "referenceName", "start", "end", "referenceBases", "alternateBases");
+
+  /**
+   * Use this PTransform for correct handling of gVCF data in DataFlow jobs that consider not only
+   * calls that have the variant but also those that match the reference at that variant position.
+   * 
+   * All variant fields will be returned.
+   * 
+   * @param input
+   * @param auth
+   * @return
+   */
+  public static PCollection<Variant> joinGvcfVariantsTransform(
+      PCollection<SearchVariantsRequest> input, GenomicsFactory.OfflineAuth auth) {
+    return joinGvcfVariants(input, auth, null);
+  }
+
+  /**
+   * Use this PTransform for correct handling of gVCF data in DataFlow jobs that consider not only
+   * calls that have the variant but also those that match the reference at that variant position.
+   * 
+   * @param input
+   * @param auth
+   * @param fields Fields to be returned by the partial response.
+   * @return
+   */
+  public static PCollection<Variant> joinGvcfVariantsTransform(
+      PCollection<SearchVariantsRequest> input, GenomicsFactory.OfflineAuth auth, String fields) {
+    for (String field : REQUIRED_FIELDS) {
+      Preconditions.checkState(fields.contains(field), "Required field missing: " + field
```

Comment:
Preconditions.checkState uses a format string to format the error message:

    Preconditions.checkState(fields.contains(field), "Required field missing: %s Add this field to the list of Variants fields returned in the partial response.", field);